clang/lib/Headers/aiev2_core.h

```diff
@@ -1566,4 +1566,300 @@ inline v16accfloat addmsc_4x8_8x4( v32float v1, v32float v2, v16accfloat acc1, v
+  __attribute__((                                                              \
+      address_space(static_cast<unsigned>(aie_dm_resource::DM_test))))
+#define __aie_dm_resource_stack                                                \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::stack))))
+#define __aie_dm_resource_a                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::a))))
+#define __aie_dm_resource_b                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::b))))
+#define __aie_dm_resource_c                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::c))))
+#define __aie_dm_resource_d                                                    \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::d))))
+#define __aie_dm_resource_ab                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ab))))
+#define __aie_dm_resource_ac                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ac))))
+#define __aie_dm_resource_ad                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::ad))))
+#define __aie_dm_resource_bc                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::bc))))
+#define __aie_dm_resource_bd                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::bd))))
+#define __aie_dm_resource_cd                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::cd))))
+#define __aie_dm_resource_TM                                                   \
+  __attribute__((address_space(static_cast<unsigned>(aie_dm_resource::TM))))
+
+template <typename T> struct aie_dm_resource_remove {
+  using type = T;
+};
+
+template <typename T, int N>
+struct aie_dm_resource_remove<T __attribute__((address_space(N)))> {
+  using type = T;
+};
+template <typename T, int N>
+struct aie_dm_resource_remove<const T __attribute__((address_space(N)))> {
+  using type = const T;
+};
+template <typename T, int N>
+struct aie_dm_resource_remove<volatile T __attribute__((address_space(N)))> {
+  using type = T;
+};
+template <typename T, int N>
+struct aie_dm_resource_remove<const volatile T
+                              __attribute__((address_space(N)))> {
+  using type = const T;
+};
+
+template <typename T>
+using aie_dm_resource_remove_t = typename aie_dm_resource_remove<T>::type;
+
+template <typename T, aie_dm_resource Resource> struct aie_dm_resource_set {
+  using type = T;
+};
+
+template <typename T> struct aie_dm_resource_set<T, aie_dm_resource::a> {
```

Comment:
It seems like these templates could be written much more compactly with some preprocessing magic.   I assume the problem here is that you can't put a template reference in an __attribute declaration?

Reply:
I hate macros, but I have to agree here. @krishnamtibrewala do you think there's something we could do to improve the readability?

Reply:
I tried ways, but none of the ways worked well for AIE API.